### PR TITLE
[receiver/githubreceiver] add skeleton for tracing GH Actions events

### PR DESCRIPTION
**Pull request by msarahan**
This is an initial skeleton of work adapted from @krzko from the discussion in #27460. It has much/most of the implementation details left out, but the general idea is to refactor what krzko had into a folder in `internal`, and plumb the configuration in with the top-level github (scraper) configuration.

I think we will want to remove the restriction that at least one scraper must be configured, since this receiver could be used with the GHA tracing and not the metrics.

CC @adrielp - not sure if this is what you had in mind for a skeleton, but I hope it is helpful. If you have more specific directions for what you'd like to see, let me know.

### CODE REVIEW

**Comment by adrielp** [comment]
@msarahan - Thanks for starting this contribution! I appreciate the quick turnaround on getting the skeleton opened up here. 

A couple things to start off with to get this PR ready for merge:

* the title needs to change, check out [title guidelines](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#title-guidelines) - something like `[receiver/githubreceiver] add GitHub tracing webhook skeleton` or similar would probably suffice
* a [changelog](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-a-changelog-entry) needs to be added. 
* looks like there's a couple workflow failures, one of which I'm unfamiliar with (tracing correctness) - will have to look into that one. but most can be solved by running the following [script locally](https://gist.github.com/adrielp/cfa0bcfbbe648d1e1d5e5ef7ac17cb0c). I put it at the same dir level as my collector-contrib fork and run it from there. 
* The metadata.yaml needs to be updated. A little below [this section](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#vendor-specific-components) in the docs is helpful. For this PR I think all that is needed is to add `development: [metrics, traces]` and re-generate. (double check that though). 

> It has much/most of the implementation details left out, but the general idea is to refactor what krzko had into a folder in internal, and plumb the configuration in with the top-level github (scraper) configuration.

Sounds right 👍🏼 

> I think we will want to remove the restriction that at least one scraper must be configured, since this receiver could be used with the GHA tracing and not the metrics.

Agreed with a slight change. until the Webhook has functionality, I don't think this should change. And when it does become functionally available, we should check to ensure at least one of metrics/traces is configured in the event that a user forgets to configure either.

**Comment by msarahan** [comment]
Thanks for the review, @adrielp. I think I fixed up most of it, except

> The metadata.yaml needs to be updated. A little below [this section](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#vendor-specific-components) in the docs is helpful. For this PR I think all that is needed is to add development: [metrics, traces] and re-generate.

I found that making this change generated code that assumed that there was a happy path for tracing, but this PR does not add that happy path.

Also, there's an error when running your prep.sh script that I've been trying to understand:

```
ERRO [linters_context/goanalysis] buildir: panic during analysis: Cannot range over: func(yield func(K, V) bool), goroutine 6175 [running]:
runtime/debug.Stack()
        runtime/debug/stack.go:26 +0x5e
github.com/golangci/golangci-lint/pkg/goanalysis.(*action).analyzeSafe.func1()
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_action.go:105 +0x5a
panic({0x1815360?, 0xc000877900?})
        runtime/panic.go:785 +0x132
honnef.co/go/tools/go/ir.(*builder).rangeStmt(0xc002465910, 0xc00266af00, 0xc00211e9c0, 0x0, {0x1d35710, 0xc00211e9c0})
        honnef.co/go/tools@v0.4.7/go/ir/builder.go:2225 +0x894
honnef.co/go/tools/go/ir.(*builder).stmt(0xc002465910, 0xc00266af00, {0x1d3a278?, 0xc00211e9c0?})
        honnef.co/go/tools@v0.4.7/go/ir/builder.go:2438 +0x20a
honnef.co/go/tools/go/ir.(*builder).stmtList(...)
        honnef.co/go/tools@v0.4.7/go/ir/builder.go:859
honnef.co/go/tools/go/ir.(*builder).stmt(0xc002465910, 0xc00266af00, {0x1d3a188?, 0xc0023965d0?})
        honnef.co/go/tools@v0.4.7/go/ir/builder.go:2396 +0x1415
honnef.co/go/tools/go/ir.(*builder).buildFunction(0xc002465910, 0xc00266af00)
        honnef.co/go/tools@v0.4.7/go/ir/builder.go:2508 +0x417
honnef.co/go/tools/go/ir.(*builder).buildFuncDecl(0xc002465910, 0xc0006e99e0, 0xc002396600)
        honnef.co/go/tools@v0.4.7/go/ir/builder.go:2545 +0x189
honnef.co/go/tools/go/ir.(*Package).build(0xc0006e99e0)
        honnef.co/go/tools@v0.4.7/go/ir/builder.go:2649 +0xb46
sync.(*Once).doSlow(0xc004c28700?, 0xc00211efc0?)
        sync/once.go:76 +0xb4
sync.(*Once).Do(...)
        sync/once.go:67
honnef.co/go/tools/go/ir.(*Package).Build(...)
        honnef.co/go/tools@v0.4.7/go/ir/builder.go:2567
honnef.co/go/tools/internal/passes/buildir.run(0xc004c28620)
        honnef.co/go/tools@v0.4.7/internal/passes/buildir/buildir.go:86 +0x18b
github.com/golangci/golangci-lint/pkg/goanalysis.(*action).analyze(0xc0037b1ed0)
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_action.go:191 +0x9cd
github.com/golangci/golangci-lint/pkg/goanalysis.(*action).analyzeSafe.func2()
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_action.go:113 +0x17
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0xc002422910, {0x1a4edad, 0x7}, 0xc00499e748)
        github.com/golangci/golangci-lint@v1.59.1/pkg/timeutils/stopwatch.go:111 +0x44
github.com/golangci/golangci-lint/pkg/goanalysis.(*action).analyzeSafe(0x0?)
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_action.go:112 +0x6e
github.com/golangci/golangci-lint/pkg/goanalysis.(*loadingPackage).analyze.func2(0xc0037b1ed0)
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_loadingpackage.go:80 +0xa5
created by github.com/golangci/golangci-lint/pkg/goanalysis.(*loadingPackage).analyze in goroutine 2743
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_loadingpackage.go:75 +0x1e9 
ERRO [linters_context/goanalysis] fact_purity: panic during analysis: interface conversion: interface {} is nil, not *buildir.IR, goroutine 6167 [running]:
runtime/debug.Stack()
        runtime/debug/stack.go:26 +0x5e
github.com/golangci/golangci-lint/pkg/goanalysis.(*action).analyzeSafe.func1()
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_action.go:105 +0x5a
panic({0x18c86e0?, 0xc002715110?})
        runtime/panic.go:785 +0x132
honnef.co/go/tools/analysis/facts/purity.purity(0xc00038ed20)
        honnef.co/go/tools@v0.4.7/analysis/facts/purity/purity.go:109 +0x299
github.com/golangci/golangci-lint/pkg/goanalysis.(*action).analyze(0xc0037c9b10)
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_action.go:191 +0x9cd
github.com/golangci/golangci-lint/pkg/goanalysis.(*action).analyzeSafe.func2()
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_action.go:113 +0x17
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0xc002422910, {0x1a98caf, 0xb}, 0xc003d0bf48)
        github.com/golangci/golangci-lint@v1.59.1/pkg/timeutils/stopwatch.go:111 +0x44
github.com/golangci/golangci-lint/pkg/goanalysis.(*action).analyzeSafe(0xc002264a80?)
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_action.go:112 +0x6e
github.com/golangci/golangci-lint/pkg/goanalysis.(*loadingPackage).analyze.func2(0xc0037c9b10)
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_loadingpackage.go:80 +0xa5
created by github.com/golangci/golangci-lint/pkg/goanalysis.(*loadingPackage).analyze in goroutine 2743
        github.com/golangci/golangci-lint@v1.59.1/pkg/goanalysis/runner_loadingpackage.go:75 +0x1e9 
ERRO [linters_context/goanalysis] typedness: panic during analysis: interface conversion: interface {} is nil, not *buildir.IR, goroutine 6176 [running]:
runtime/debug.Stack()
```

It seems like the "real" error is `Cannot range over: func(yield func(K, V) bool)`, and that `typedness: panic during analysis: interface conversion: interface {} is nil, not *buildir.IR` might be "aftermath", but my go is rusty. I'll keep working on it, but any insight you can add would be much appreciated.

**Comment by adrielp** [comment]
> The metadata.yaml needs to be updated. A little below [this section](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#vendor-specific-components) in the docs is helpful. For this PR I think all that is needed is to add development: [metrics, traces] and re-generate.
> I found that making this change generated code that assumed that there was a happy path for tracing, but this PR does not add that happy path.

I see that component lifecycle tests are generated for testing and fail with:

```bash
│                                                                                                                                                                     │
│=== Failed                                                                                                                                                           │
│=== FAIL: . TestComponentLifecycle/traces-shutdown (0.00s)                                                                                                           │
│    generated_component_test.go:59:                                                                                                                                  │
│                Error Trace:    /Users/adrielperkins/dev/trash/opentelemetry-collector-contrib/receiver/githubreceiver/generated_component_test.go:59                │
│                Error:          Received unexpected error:                                                                                                           │
│                                telemetry type is not supported                                                                                                      │
│                Test:           TestComponentLifecycle/traces-shutdown                                                                                               │
│                                                                                                                                                                     │
│=== FAIL: . TestComponentLifecycle/traces-lifecycle (0.00s)                                                                                                          │
│    generated_component_test.go:65:                                                                                                                                  │
│                Error Trace:    /Users/adrielperkins/dev/trash/opentelemetry-collector-contrib/receiver/githubreceiver/generated_component_test.go:65                │
│                Error:          Received unexpected error:                                                                                                           │
│                                telemetry type is not supported                                                                                                      │
│                Test:           TestComponentLifecycle/traces-lifecycle                                                                                              │
│                                                                                                                                                                     │
│=== FAIL: . TestComponentLifecycle (2.01s)    
```

I need to do the in-depth code review beyond the core fundamentals already mentioned. I'm thinking we should be able to add a tiny bit more to make this pass where the receiver with traces configured can run no-op.

> 
> Also, there's an error when running your prep.sh script that I've been trying to understand:
> 
> ```
> ERRO [linters_context/goanalysis] buildir: panic during analysis: Cannot range over: func(yield func(K, V) bool), goroutine 6175 [running]:
> runtime/debug.Stack()
> .........
> ```

So looks like this is a bug in golanglint-ci. I went in on your branch and checked out running it against other receivers. I ended up doing the following:

```bash
cd opentelemetry-collector-contrib/internal/tools
go get -u github.com/golangci/golangci-lint && make tidy
cd - && cd receiver/githubreceiver
../../.tools/golangci-lint --config ../../.golangci.yml run
```

Doing so updated golanglint-ci from `v1.59.1` to `v1.60.3` which had no issues.

File: go.mod
Comment (adrielp) [inline go.mod:1]: I'm curious why changes to `go.mod` at the root is showing up in this pull request?

File: cmd/otelcontribcol/go.mod
Comment (adrielp) [inline cmd/otelcontribcol/go.mod:1]: I'm curious why changes to `go.mod` in cmd/otelcontribcol is showing up in this pull request.

File: receiver/githubreceiver/config_test.go [hunk 1]
```diff
@@ -11,13 +11,16 @@ import (
...
+	"go.opentelemetry.io/collector/config/confighttp"
 	"go.opentelemetry.io/collector/confmap"
 	"go.opentelemetry.io/collector/otelcol/otelcoltest"
 	"go.opentelemetry.io/collector/receiver/scraperhelper"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/githubreceiver/internal"
 	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/githubreceiver/internal/metadata"
 	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/githubreceiver/internal/scraper/githubscraper"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/githubreceiver/internal/webhook"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/webhookeventreceiver"
```
Comment (adrielp) [inline receiver/githubreceiver/config_test.go:23]: I don't think we should be importing the webhookevent receiver component as part of our tests here and instead should follow the pattern the webhookevent receiver uses to test itself.

File: receiver/githubreceiver/config_test.go [hunk 2]
```diff
@@ -40,6 +43,14 @@ func TestLoadConfig(t *testing.T) {
 	defaultConfigGitHubScraper.(*Config).Scrapers = map[string]internal.Config{
 		githubscraper.TypeStr: (&githubscraper.Factory{}).CreateDefaultConfig(),
 	}
+	defaultConfigGitHubScraper.(*Config).WebhookConfig = webhook.Config{
```
Comment (adrielp) [inline receiver/githubreceiver/config_test.go:46]: Webhook should be separate from `defaultConfigGitHubScraper`.

File: receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go [hunk 3]
```diff
@@ -262,10 +262,10 @@ func TestGetAge(t *testing.T) {
 
 	for _, tc := range testCases {
 		t.Run(tc.desc, func(t *testing.T) {
-			min := time.Now()
-			max := min.Add(tc.hrsAdd).Add(tc.minsAdd)
+			_min := time.Now()
```
Comment (adrielp) [inline receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go:265]: what was the purpose in making these changes?

File: receiver/githubreceiver/internal/webhook/config.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	"go.uber.org/multierr"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/webhookeventreceiver"
+)
+
+var errMissingEndpointFromConfig = errors.New("missing receiver server endpoint from config")
+
+// Config defines configuration for GitHub Actions receiver
+type Config struct {
+	WebhookReceiver   webhookeventreceiver.Config `mapstructure:",squash"`             // squash ensures fields are correctly decoded in embedded struct
```
Comment (adrielp) [inline receiver/githubreceiver/internal/webhook/config.go:19]: see my comment about use of the webhookevent receiver as a library within this component.

File: receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go (hunk 3 above)
Comment (msarahan) [reply receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go:265]: One of the linters raised an issue with the code as it was with `min` and `max` clobbering the built-in functions. I was mostly just trying to fix everything that the linter mentioned, but I'm fine to remove this change to reduce the surface area of the PR.

File: go.mod
Comment (msarahan) [reply go.mod:1]: I believe it may have to do with me doing a `go mod install` for a linter that was used in the script you recommended to me. It definitely does not belong in this PR, and thanks for pointing it out.

File: receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go (hunk 3 above)
Comment (adrielp) [reply receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go:265]: interesting. i wonder if it was a recent update to the linter since it hadn't been caught previously. fine enough to leave in

File: receiver/githubreceiver/config_test.go (hunk 2 above)
Comment (msarahan) [reply receiver/githubreceiver/config_test.go]: I renamed this `defaultConfigGithubReceiver to make it more clear that it is the top-level configuration struct. I also renamed `WebhookConfig` to `Tracing`.

File: receiver/githubreceiver/config_test.go (hunk 1 above)
Comment (msarahan) [reply receiver/githubreceiver/config_test.go:23]: I'm not sure what you mean here. I am using the webhookreceiver to try to avoid reimplementing the webhook functionality. I'm interpreting your request as "please don't use code from other receivers" - which I'm fine to respect. One thing that I've tried to interpret this is to use anonymous structs that have the same fields as the expected webhook configuration struct. Struct stuff doesn't really work that way (much to my Python-headed duck-typing consternation), so the next nearest thing is an interface with getter-type functions. Am I on the right track, or do you mean something else?

One thing I'm a little hung up on is that the `Validate()` method for the webhookreceiver's config is being called somehow, and I can't figure out when or how. Perhaps this is one reason why you're avoiding using the webhookreceiver.Config struct. I have a debugger breakpoint on the only line where that can be added to our error collection, and the breakpoint is never hit. Nevertheless, the tests fail with:

```
=== RUN   TestLoadConfig
    config_test.go:36: 
                Error Trace:    /home/msarahan/code/opentelemetry-collector-contrib/receiver/githubreceiver/config_test.go:36
                Error:          Received unexpected error:
                                receivers::github/customname: missing receiver server endpoint from config
                Test:           TestLoadConfig
--- FAIL: TestLoadConfig (0.00s)
=== RUN   TestLoadConfig_Traces
    config_test.go:89: 
                Error Trace:    /home/msarahan/code/opentelemetry-collector-contrib/receiver/githubreceiver/config_test.go:89
                Error:          Received unexpected error:
                                receivers::github/customname: missing receiver server endpoint from config
                Test:           TestLoadConfig_Traces
```

File: receiver/githubreceiver/metadata.yaml [hunk 5]
```diff
@@ -5,7 +5,7 @@ sem_conv_version: 1.27.0
 status:
   class: receiver
   stability:
-    alpha: [metrics]
+    alpha: [metrics,tracing]
```
Comment (Elfo404) [inline receiver/githubreceiver/metadata.yaml:8]: should this be `traces` according to https://github.com/open-telemetry/opentelemetry-collector/blob/main/cmd/mdatagen/metadata-schema.yaml?

Comment (adrielp) [reply receiver/githubreceiver/metadata.yaml:8]: correct!

File: receiver/githubreceiver/config.go [hunk 6]
```diff
@@ -24,6 +25,7 @@ type Config struct {
 	scraperhelper.ControllerConfig `mapstructure:",squash"`
 	Scrapers                       map[string]internal.Config `mapstructure:"scrapers"`
 	metadata.MetricsBuilderConfig  `mapstructure:",squash"`
+	Traces                         traces.Config `mapstructure:"traces"`
```
Comment (adrielp) [inline receiver/githubreceiver/config.go:28]: @Elfo404 - This is what I was talking about. Could you give some purview on the complications or ease of backend logic that is able so separate events that are built into traces vs events that could be built into metrics of just sent as raw events? 

My thinking here, and this could come in a subsequent iteration, is that it'd be simpler to only require the configuration of one GitHub app to sent all data, instead of requiring the configuration of multiple to point to different hooks that process logs and traces.

File: receiver/githubreceiver/config_test.go (hunk 1 above)
Comment (adrielp) [reply receiver/githubreceiver/config_test.go]: > I am using the webhookreceiver to try to avoid reimplementing the webhook functionality. I'm interpreting your request as "please don't use code from other receivers" - which I'm fine to respect. One thing that I've tried to interpret this is to use anonymous structs that have the same fields as the expected webhook configuration struct. 

I think my suggestion is to re-implement it's functionality so that this component is entirely decoupled from any changes to the WebHook receiver component. This removes a dependency that can change separately from this code. 

@crobert-1, @TylerHelmuth, @andrzej-stencel - I'm curious to hear your thoughts here and get your opinion.

File: receiver/githubreceiver/internal/traces/README.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
+# GitHub Actions Receiver
+
+<!-- status autogenerated section -->
```
Comment (adrielp) [inline receiver/githubreceiver/internal/traces/README.md:3]: This whole auto generated section can be deleted. 

I think most of the information in this readme should be in the readme at the root of the receiver, and if necessary, can link to this as an internal document after adjustement.

File: receiver/githubreceiver/internal/traces/config.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	Secret            string                      `mapstructure:"secret"`              // GitHub traces signature. Default is empty
+	CustomServiceName string                      `mapstructure:"custom_service_name"` // custom service name. Default is empty
+	ServiceNamePrefix string                      `mapstructure:"service_name_prefix"` // service name prefix. Default is empty
+	ServiceNameSuffix string                      `mapstructure:"service_name_suffix"` // service name suffix. Default is empty
+}
+
+var _ component.Config = (*Config)(nil)
+
+// Validate checks the receiver configuration is valid
+func (cfg *Config) Validate() error {
```
Comment (adrielp) [inline receiver/githubreceiver/internal/traces/config.go:29]: TODO: revisit this after the discussion of using the webhook receiver as part of this or re-implementing similar code.

File: receiver/githubreceiver/metadata.yaml (hunk 5 above)
Comment (adrielp) [inline receiver/githubreceiver/metadata.yaml:8]:
```suggestion
    alpha: [metrics, traces]
```

File: receiver/githubreceiver/config.go (hunk 6 above)
Comment (Elfo404) [reply receiver/githubreceiver/config.go:28]: yeah, I was looking at a previous revision which called it Webhook.

I think keeping it simple from a config POV is better (i.e. reverting back to calling this `webhook`). Reason being you wouldn't have to manage multiple GH apps. (which if needed for other reasons you can still do by having different instances of the receiver with different config, for instance because of rate limits). Then we can add different config sections to customize the behavior of how traces or metrics are generated.

Internally is pretty straightforward to differentiate between events; take a look at what we did in https://github.com/grafana/grafana-ci-otel-collector/blob/main/receiver/githubactionsreceiver/receiver.go#L229, where we differentiate on the event type, and generate metrics on non-skipped `workflow_job` events, and after generate traces and logs on completed `workflow_job` and `workflow_run` events (https://github.com/grafana/grafana-ci-otel-collector/blob/main/receiver/githubactionsreceiver/receiver.go#L257-L293).

it's a bit of spaghetti code, but the metrics bit was thrown together in very little time. I'm sure that with some routing logic we can keep the implementation clean and not require users to define different webhooks.

TLDR; my thought is that the configuration should be independent on how the event handling is implemented, the fact that traces are generated thanks to the webhook is an implementation detail. WDYT?

**Comment by msarahan** [comment]
My work has taken a turn away from this, and I'm afraid I probably can't continue here. Would anyone like to pick up this work, or shall I close this PR?

**Comment by adrielp** [comment]
> My work has taken a turn away from this, and I'm afraid I probably can't continue here. Would anyone like to pick up this work, or shall I close this PR?

Firstly, thanks for the work on this @msarahan. Secondly thanks for letting us know. I'll try to pick up this work from here. I know contributions can take time, but if you ever have bandwidth again, feel free to contribute, always happy to see new folks contributing!
